nixos/modules/services/monitoring/telegraf.nix

```diff
@@ -0,0 +1,71 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.telegraf;
+
+  configFile = pkgs.runCommand "config.toml" {
+    buildInputs = [ pkgs.remarshal ];
+  } ''
+    remarshal -if json -of toml \
+      < ${pkgs.writeText "config.json" (builtins.toJSON cfg.extraConfig)} \
+      > $out
+  '';
+in
+{
+
+  ###### interface
+
+  options = {
+
+    services.telegraf = {
+
+      enable = mkEnableOption "telegraf server";
+
+      package = mkOption {
+        default = pkgs.telegraf;
+        defaultText = "pkgs.telegraf";
+        description = "Which telegraf derivation to use";
+        type = types.package;
+      };
+
+      extraConfig = mkOption {
+        default = {};
+        description = "Extra configuration options for telegraf";
+        type = types.attrs;
+      };
+    };
+
+  };
+
+
+  ###### implementation
+
+  config = mkIf config.services.telegraf.enable {
+
+    systemd.services.telegraf = {
+      description = "Telegraf Agent";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-interfaces.target" ];
```

Comment:
@joachifm a follow up to #18319 is on my to-do to get rid of `network-interfaces.target` since so many targets tend to confuse everyone. [upstream network targets](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/) should really be enough for ordinary services.
feel free to assist :wink: